cmd/juju/interact/pollster.go

```diff
@@ -0,0 +1,515 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package interact
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"html/template"
+	"io"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/jsonschema"
+	"github.com/juju/utils/set"
+)
+
+// Pollster is used to ask multiple questions of the user using a standard
+// formatting.
+type Pollster struct {
+	scanner *bufio.Scanner
+	out     io.Writer
+	errOut  io.Writer
+}
+
+// New returns a Pollster that wraps the given reader and writer.
+func New(in io.Reader, out, errOut io.Writer) *Pollster {
+	return &Pollster{
+		scanner: bufio.NewScanner(in),
+		out:     out,
+		errOut:  errOut,
+	}
+}
+
+// List contains the information necessary to ask the user to select one item
+// from a list of options.
+type List struct {
+	Singular string
+	Plural   string
+	Options  []string
+	Default  string
+}
+
+// MultiList contains the information necessary to ask the user to select form a list
+// of options.
+type MultiList struct {
+	Singular string
+	Plural   string
+	Options  []string
+	Default  []string
+}
+
+var listTmpl = template.Must(template.New("").Funcs(map[string]interface{}{"title": strings.Title}).Parse(`
+{{title .Plural}}
+{{range .Options}}  {{.}}
+{{end}}
+`[1:]))
+
+var selectTmpl = template.Must(template.New("").Parse(`
+Select {{.Singular}}{{if .Default}} [{{.Default}}]{{end}}: `[1:]))
+
+// Select queries the user to select from the give list of options.
+func (p *Pollster) Select(l List) (string, error) {
+	return p.SelectVerify(l, VerifyOptions(l.Singular, l.Options, l.Default != ""))
+}
+
+// SelectVerify queries the user to select from the give list of options, using
+// the custom verifier listed.
+func (p *Pollster) SelectVerify(l List, verify VerifyFunc) (string, error) {
+	if err := listTmpl.Execute(p.out, l); err != nil {
+		return "", err
+	}
+
+	question, err := sprint(selectTmpl, l)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	val, err := QueryVerify(question, p.scanner, p.out, p.errOut, verify)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	if val == "" {
+		return l.Default, nil
+	}
+	return val, nil
+}
+
+var multiSelectTmpl = template.Must(template.New("").Funcs(
+	map[string]interface{}{"join": strings.Join}).Parse(`
+Select one or more {{.Plural}} separated by commas{{if .Default}} [{{join .Default ", "}}]{{end}}: `[1:]))
+
+// MultiSelect queries the user to select one more answers from the give list of
```

Comment:
Not sure about "extra", but when you enter text, it is split on commas.  There's no actual sanity check that the options themselves don't contain commas (which would make them unusable)... I'll add that.